From bcff0f3e52ec5ebd0c642be6f37cadbe89613cb2 Mon Sep 17 00:00:00 2001
From: Claudio Bley <claudio.bley@tweag.io>
Date: Tue, 3 Feb 2026 13:24:49 +0100
Subject: [PATCH 15/22] Use `max_retries` setting

Retry 5 times on network / RE retryable errors. This is the same value Bazel uses.
---
 app/buck2_re_configuration/src/lib.rs      |  6 ++++--
 remote_execution/oss/re_grpc/src/client.rs | 17 ++++++++---------
 2 files changed, 12 insertions(+), 11 deletions(-)

diff --git a/app/buck2_re_configuration/src/lib.rs b/app/buck2_re_configuration/src/lib.rs
index 49e2bebc3b..7e7ec6baa0 100644
--- a/app/buck2_re_configuration/src/lib.rs
+++ b/app/buck2_re_configuration/src/lib.rs
@@ -19,6 +19,8 @@ use buck2_core::rollout_percentage::RolloutPercentage;
 
 static BUCK2_RE_CLIENT_CFG_SECTION: &str = "buck2_re_client";
 
+const DEFAULT_MAX_RETRIES: usize = 5;
+
 /// We put functions here that both things need to implement for code that isn't gated behind a
 /// fbcode_build or not(fbcode_build)
 pub trait RemoteExecutionStaticMetadataImpl: Sized {
@@ -449,7 +451,7 @@ pub struct Buck2OssReConfiguration {
     pub grpc_keepalive_timeout_secs: Option<u64>,
     /// Whether to send HTTP/2 pings when connection is idle.
     pub grpc_keepalive_while_idle: Option<bool>,
-    /// Maximum retries for network requests.
+    /// Maximum retries for RPC requests. Defaults to 5.
     pub max_retries: usize,
     /// Timeout for RPC requests in seconds. Defaults to 60s.
     pub rpc_timeout: u64,
@@ -575,7 +577,7 @@ impl Buck2OssReConfiguration {
                     section: BUCK2_RE_CLIENT_CFG_SECTION,
                     property: "max_retries",
                 })?
-                .unwrap_or(0),
+                .unwrap_or(DEFAULT_MAX_RETRIES),
             rpc_timeout: legacy_config
                 .parse(BuckconfigKeyRef {
                     section: BUCK2_RE_CLIENT_CFG_SECTION,
diff --git a/remote_execution/oss/re_grpc/src/client.rs b/remote_execution/oss/re_grpc/src/client.rs
index cb0c93874b..a6d66b9ad8 100644
--- a/remote_execution/oss/re_grpc/src/client.rs
+++ b/remote_execution/oss/re_grpc/src/client.rs
@@ -104,7 +104,6 @@ use crate::retry::retry;
 use crate::stats::CountingConnector;
 
 const DEFAULT_MAX_TOTAL_BATCH_SIZE: usize = 4 * 1000 * 1000;
-const MAX_RETRIES: usize = 5;
 const INITIAL_DELAY: Duration = Duration::from_millis(100);
 const MAX_DELAY: Duration = Duration::from_secs(10);
 
@@ -720,7 +719,7 @@ impl REClient {
                     ttl: 0,
                 })
             },
-            MAX_RETRIES,
+            self.runtime_opts.max_retries,
             INITIAL_DELAY,
             MAX_DELAY,
             false,
@@ -758,7 +757,7 @@ impl REClient {
                     ttl_seconds: 0,
                 })
             },
-            MAX_RETRIES,
+            self.runtime_opts.max_retries,
             INITIAL_DELAY,
             MAX_DELAY,
             false,
@@ -802,7 +801,7 @@ impl REClient {
                     .into_inner();
                 Ok(stream)
             },
-            MAX_RETRIES,
+            self.runtime_opts.max_retries,
             INITIAL_DELAY,
             MAX_DELAY,
             true,
@@ -945,7 +944,7 @@ impl REClient {
                             Ok(resp.into_inner())
                         }
                     },
-                    MAX_RETRIES,
+                    self.runtime_opts.max_retries,
                     INITIAL_DELAY,
                     MAX_DELAY,
                     false,
@@ -970,7 +969,7 @@ impl REClient {
                             Ok(resp.into_inner())
                         }
                     },
-                    MAX_RETRIES,
+                    self.runtime_opts.max_retries,
                     INITIAL_DELAY,
                     MAX_DELAY,
                     false,
@@ -1038,7 +1037,7 @@ impl REClient {
                                 .into_inner())
                         }
                     },
-                    MAX_RETRIES,
+                    self.runtime_opts.max_retries,
                     INITIAL_DELAY,
                     MAX_DELAY,
                     false,
@@ -1077,7 +1076,7 @@ impl REClient {
                                     >)
                             }
                         },
-                        MAX_RETRIES,
+                        self.runtime_opts.max_retries,
                         INITIAL_DELAY,
                         MAX_DELAY,
                         false,
@@ -1144,7 +1143,7 @@ impl REClient {
                                 .context("Failed to request what blobs are not present on remote")
                         }
                     },
-                    MAX_RETRIES,
+                    self.runtime_opts.max_retries,
                     INITIAL_DELAY,
                     MAX_DELAY,
                     false,
-- 
2.51.2

